packages/flutter_tools/lib/src/run.dart

```diff
@@ -288,9 +293,78 @@ class RunAndStayResident {
     });
   }
 
+  DevFS devFS;
+
+  Future<Null> _updateDevFS() async {
+    if (devFS == null) {
+      devFS = new DevFS(Directory.current, observatory);
+
+      try {
+        await devFS.init();
+      } catch (error) {
+        devFS = null;
+        printError('Error initing DevFS: $error');
+        return null;
+      }
+    }
+
+    // Send the root and lib directories.
+    Directory directory = Directory.current;
+    _sendFiles(directory, '', _dartFiles(directory.listSync()));
+
+    directory = new Directory('lib');
+    _sendFiles(directory, 'lib', _dartFiles(directory.listSync(recursive: true, followLinks: false)));
+
+    // Send the packages.
+    if (FileSystemEntity.isFileSync(kPackagesFileName)) {
+      PackageMap packageMap = new PackageMap(kPackagesFileName);
+
+      for (String packageName in packageMap.map.keys) {
+        Uri uri = packageMap.map[packageName];
+        // Ignore self-references.
+        if (uri.toString() == 'lib/')
+          continue;
+        Directory directory = new Directory.fromUri(uri);
+        if (directory.existsSync()) {
+          _sendFiles(
+            directory,
+            'packages/$packageName',
+            _dartFiles(directory.listSync(recursive: true, followLinks: false))
+          );
+        }
+      }
+    }
+
+    try {
+      await devFS.flush();
+    } catch (error) {
+      printError('Error sending DevFS files: $error');
+    }
+  }
+
+  void _sendFiles(Directory base, String prefix, List<File> files) {
+    String basePath = base.path;
+
+    for (File file in files) {
+      String devPath = file.path.substring(basePath.length);
+      if (devPath.startsWith('/'))
+        devPath = devPath.substring(1);
+      devFS.stageFile(prefix.isEmpty ? devPath : '$prefix/$devPath', file);
+    }
+  }
+
+  List<File> _dartFiles(List<FileSystemEntity> entities) {
+    return new List<File>.from(entities
+      .where((FileSystemEntity entity) => entity is File)
+      .where((File file) => file.path.endsWith('.dart')));
```

Comment:
These two `.where` could be combined into one: `entity is File && file.path.endsWith('.dart')`